2 Niagara Fundamentals Exercise - Multiplexer/multiplexer-rt/src/com/tridiumuniversity/multiplexer/BMultiplexer.java

```diff
@@ -0,0 +1,284 @@
+    private void calcSInputCount()
+    {
+        sInputCount = 0; // reset the old count
+        sInputs.clear(); // clear the old "s" input list
+        SlotCursor<Property> propertyCursor = getProperties(); // create the cursor
+
+        while (propertyCursor.next()) // iterate through the properties
+        {
+            Property property = propertyCursor.property(); // get property object
+            try
+            {
+                BBoolean value = (BBoolean) propertyCursor.get(); // get property value. If the property isn't a BBoolean, skip
+
+                if (property.getName().startsWith("s")) {
+                    sInputCount += 1; // add one to the count
+
+                    sInputs.add(property);
+                }
+            }
+            catch (Exception e)
+            {
+            }
+        }
+        log.fine("Counted Ss: " + sInputCount); // FOR SOME REASON THIS LOGS IN THE WB CONSOLE WHEN SWITCHING VIEWS
+    }
+
+
+    /* Calculates the effective input step. The input step decides how many sInputs are needed/used */
+    private void calcInputStep()
+    {
+        /* -1 in the loop so the iteration number matches with the index number of the array.
+        I don't know if this is how it's normally done. It's the first time I've ever wanted to count backwards in a loop */
+        for (int i = inputSteps.length - 1; i>-1; i--)
+        //
+        {
+            if (inInputCount >= inputSteps[i])
+            {
+                effInputStep = i;
+                log.fine("Effective input step: " + effInputStep);
+                break;
+            }
+        }
+    }
+
+
+    /* Calculates the switchSum based on the combination of switches currently set to "true" */
+    private void calcSwitchSum()
+    {
+        switchSum = 0; // reset the switchSum
+
+        /* process all the "in" inputs */
+        for (int i = 0; i<sInputs.size(); i++)
+        {
+            if (((BBoolean) sInputs.get(i).getDefaultValue()).getBoolean()) // get input as boolean. If true, add a numeric value to the switchSum
+            {
+                String propertyName = sInputs.get(i).getName(); // get the property's name
+                int sNumber = Integer.parseInt(propertyName.replace("s", "")); // capture the number from the property's name
+
+                int switchValue = switchValues[sNumber]; // figure out the switch's numeric value based on the switch number
+
+                if (sNumber <= effInputStep - 1) // if the switch number is an "active" switch...
+                {
+                    switchSum += switchValue; // apply the numeric value towards the sum
+                }
+            }
+        }
+
+        setUsableSwitches(effInputStep);
+        setSwitchTotal(switchSum);
+    }
+
+
+    /* updates the out property */
+    private void updateOut()
+    {
+        log.info("Elevated Property: " + "in" + switchSum); // FOR SOME REASON THIS LOGS IN THE WB CONSOLE WHEN SWITCHING VIEWS
```

Comment:
To address your comment: we haven't talked about this in class yet, but this happens because the callbacks, like `added`, `changed`, etc. sometimes run in both the Workbench JVM and in the station JVM. This is generally something we want to avoid, we just haven't covered how yet. It boils down to adding the following line to the top of your callback methods:

```java
if (!isRunning()) return;
```